pipenv/core.py

```diff
@@ -1511,7 +1511,8 @@ def pip_install(
         trusted_hosts=trusted_hosts,
         pypi_mirror=pypi_mirror,
     )
-    if not search_all_sources and requirement.index in sources:
+    source_names = {src.get("name") for src in sources}
+    if not search_all_sources and requirement.index in source_names:
```

Comment:
I think this change might have broken some tests. The proposed change only shows that the previous code would never find `requirement.index` in `sources` because the types didn't match. `str` and `List[TSource]`

Reply:
@aronbierbaum This patch fixes the other test that does fail:
```
diff --git a/tests/integration/test_lock.py b/tests/integration/test_lock.py
index 28e7f21f..d571a8d0 100644
--- a/tests/integration/test_lock.py
+++ b/tests/integration/test_lock.py
@@ -335,10 +335,11 @@ name = "testpypi"

 [packages]
 pipenv-test-private-package = {version = "*", index = "testpypi"}
+pipenv-test-public-package = {version = "*", index = "pypi"}
 requests = "*"
             """.strip()
             f.write(contents)
```

Reply:
The reason that is important is because it doesn't search all sources and `pipenv-test-private-package` is in testpypi but requires `pipenv-test-public-package` is in `pypi`.   When we initially switched to index restricted packages, there were some users that had to do similar things for their Pipfile index restrictions to work.